### PR TITLE
Refactor macOS CI so that tests run in parallel

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Refactor macOS CI so that tests run in parallel 
  - Also:
    - Fix bootstrap issue where curl was failing for a month (so obviously not needed)
    - Change Windows test stages to use PowerShell Core (where we can) to avoid compat issues

## PR Context

This allows running all tests in about the same about of clock time.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: tools/ci.psm1 [hunk 1]
```diff
@@ -751,74 +793,88 @@ function Invoke-LinuxTests
...
-            }
+    # If the tests did not pass, throw the reason why
+    if ( $result -eq "FAIL" )
+    {
+        Write-Warning "Tests failed. See the issue below."
+        Throw $resultError
+    }
+    else
+    {
+        Write-Verbose "Tests did not fail! Nice job!"
```
Comment (RDIL) [inline tools/ci.psm1]:
```suggestion
        Write-Verbose "Tests passed! Nice job!"
```

**Review by RDIL — APPROVED** [review]
Looks good so far, just saw this and thought it could be simplified.

File: tools/WindowsCI.psm1 [hunk 2]
```diff
@@ -0,0 +1,74 @@
+if($PSVersionTable.PSEdition -ne 'Desktop')
```
Comment (TravisEz13) [inline tools/WindowsCI.psm1]: This is code that other that this imperative code to protect against loading in PowerShell core was just moved from CI.psm1

File: .vsts-ci/mac.yml [hunk 3]
```diff
@@ -35,43 +35,64 @@ resources:
...
+        xunit/**/*
+      downloadPath: '$(System.ArtifactsDirectory)'
 
-  - powershell: |
-      Import-Module ./tools/ci.psm1
-      Invoke-Bootstrap-Stage
-    displayName: Bootstrap
-    condition: succeeded()
+  - pwsh: |
+      dir "$(System.ArtifactsDirectory)\*" -Recurse
```
Comment (adityapatwardhan) [inline .vsts-ci/mac.yml]: `dir` -> `Get-ChildItem`

File: .vsts-ci/templates/nix-test.yml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  - task: DownloadBuildArtifacts@0
+    displayName: 'Download build artifacts'
+    inputs:
+      downloadType: specific
+      itemPattern: |
+        build/**/*
+      downloadPath: '$(System.ArtifactsDirectory)'
+
+  - pwsh: |
+      dir "$(System.ArtifactsDirectory)\*" -Recurse
```
Comment (adityapatwardhan) [inline .vsts-ci/templates/nix-test.yml]: `dir` -> `Get-ChildItem`

File: .vsts-ci/templates/nix-test.yml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      destinationFolder: '$(System.ArtifactsDirectory)/bins'
+
+  - bash: |
+      find "$(System.ArtifactsDirectory)/bins" -type d -exec chmod +rwx {} \;
+      find "$(System.ArtifactsDirectory)/bins" -type f -exec chmod +rw {} \;
+    displayName: 'Fix permissions'
+    continueOnError: true
+
+  - pwsh: |
+      dir "$(System.ArtifactsDirectory)\bins\*" -Recurse -ErrorAction SilentlyContinue
```
Comment (adityapatwardhan) [inline .vsts-ci/templates/nix-test.yml]: `dir` -> `Get-ChildItem`

File: .vsts-ci/templates/nix-test.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    displayName: 'Capture extracted build zip'
+    continueOnError: true
+
+  - pwsh: |
+      Import-Module .\tools\ci.psm1
+      Restore-PSOptions -PSOptionsPath '$(System.ArtifactsDirectory)\build\psoptions.json'
+      $options = (Get-PSOptions)
+      $rootPath = '$(System.ArtifactsDirectory)\bins'
+      $originalRootPath = Split-Path -path $options.Output
+      Write-Host "orp- '$originalRootPath'"
```
Comment (adityapatwardhan) [inline .vsts-ci/templates/nix-test.yml]: Remove debug comments

**Review by adityapatwardhan — APPROVED** [review]

File: .vsts-ci/mac.yml (hunk 3 above)
Comment (TravisEz13) [inline .vsts-ci/mac.yml]:
```suggestion
      Get-ChildItem "$(System.ArtifactsDirectory)\*" -Recurse
```

File: .vsts-ci/templates/nix-test.yml (hunk 4 above)
Comment (TravisEz13) [reply .vsts-ci/templates/nix-test.yml]:
```suggestion
      Get-ChildItem "$(System.ArtifactsDirectory)\*" -Recurse
```

File: .vsts-ci/templates/nix-test.yml (hunk 5 above)
Comment (TravisEz13) [reply .vsts-ci/templates/nix-test.yml]:
```suggestion
      Get-ChildItem "$(System.ArtifactsDirectory)\bins\*" -Recurse -ErrorAction SilentlyContinue
```

File: .vsts-ci/templates/windows-test.yml [hunk 7]
```diff
@@ -27,19 +27,19 @@ jobs:
         build/**/*
       downloadPath: '$(System.ArtifactsDirectory)'
 
-  - powershell: |
+  - pwsh: |
       dir "$(System.ArtifactsDirectory)\*" -Recurse
```
Comment (TravisEz13) [inline .vsts-ci/templates/windows-test.yml]:
```suggestion
      Get-ChildItem "$(System.ArtifactsDirectory)\*" -Recurse
```

**Pull request merged**